frontend/components/CookieBar.test.js

```diff
@@ -0,0 +1,30 @@
+import { render, screen, fireEvent } from "@testing-library/react";
+import { getCookieConsentValue } from "react-cookie-consent";
+import jestConfig from "../jest.config";
+import CookieBar from "./CookieBar";
+
+describe("Cookiebar", () => {
+  const onAcceptMock = jest.fn();
```

Comment:
As I mentioned in the feedback for CookieBar.js, the useEffect needs to actually call the `onAccept` function. The third test ("should handle accept") is currently incorrect because you're creating the `onAcceptMock` once and then reusing that same mock every time.

On line 21, you assert that the mock was called once, however nothing in that test would actually cause the mock to be called. If you remove the second test ("disappears when you click a button"), the third test will fail. That's because it's the second test that causes the mock to be called.

I think there are a couple of ways we can improve this:

### 1. Remove the third test

Remove the third test, and check the mock after clicking the button:

```js
  it("disappears when you click a button", async () => {
    fireEvent.click(screen.getByRole("button", { name: /Accept cookies/i }));

    expect(screen.queryByText("Accept cookies")).not.toBeInTheDocument();
    expect(onAcceptMock.mock.calls.length).toBe(1);
  });
```

This is a pretty good choice because it asserts that clicking the button will remove the cookie bar and that doing so will call [the `CookieConsent` `onAccept` function](https://github.com/ZEnMo/Holon-webapp/blob/843fa702ceca5754a89959f9d379e5db80963f10/frontend/components/CookieBar.js#L26) when the user accepts.

### 2. Remove the `useEffect`

Is the `useEffect` necessary? Looking through the source code for the library, it seems as if [the cookie bar will be hidden by the library](https://github.com/Mastermindzh/react-cookie-consent/blob/a9612771f2a0d51c066be9e26936464e89e2cf44/src/CookieConsent.tsx#L18-L24) if the user has already made a choice. If we did remove it, we'd still need a way to start Google Analytics if they have previously accepted the agreement.

Perhaps we could move this to a separate component?

```js
import { useEffect } from "react";

import{ getCookieConsentValue } from "react-cookie-consent";
import { initGA } from "../util/gtag";

function Analytics() {
  useEffect(() => {
    getCookieConsentValue() === "true") && initGA();
  }, []);

  render <CookieBar onAccept={initGA} />;
}
```

It would probably be hard to test this, but it has the advantage that the "index.js" page doesn't need to define `onAccept`, while the `CookieBar` gets simpler. Another upside is that `Analytics` can be included in the other pages without them having to define their own version of `onAccept`.

Alternatively, if we want the cookie consent (and Analytics) to appear on any page, we could move it to pages/_app.js:

```js
import { useEffect } from "react";
import PropTypes from "prop-types";

import { getCookieConsentValue } from "react-cookie-consent";
import { initGA } from "../util/gtag";

import "@fontsource/inter/variable.css";

function MyApp({ Component, pageProps }) {
  useEffect(() => {
    getCookieConsentValue() === "true") && initGA();
  }, []);

  return (
    <>
      <Component {...pageProps} />
      <CookieBar onAccept={initGA} />
    </>
  );
}

export default MyApp;

MyApp.propTypes = {
  Component: PropTypes.func,
  pageProps: PropTypes.object,
};
```

What do you think? I'm open to other ideas, this is just a suggestion!

Reply:
Thank you so much for this! I like the option of adding the cookie consent to every page. This ensures we don't forget to add it to pages if we expand the tool in the future. I've added this to the MyApp component.